palace/utils/geodata.cpp

```diff
@@ -1571,7 +1615,344 @@ std::unique_ptr<mfem::Mesh> LoadMesh(const std::string &path, bool remove_curvat
+        element_types.has_pyramids)
+    {
+      MFEM_VERIFY(!mesh->Nonconforming(),
+                  "Mesh element splitting is not supported for nonconforming meshes!");
+      MFEM_VERIFY(!element_types.has_prisms && !element_types.has_pyramids,
+                  "Splitting mesh elements to hexahedra only supports simplex elements "
+                  "(tetrahedra) for now!");
+      if (preserve_curvature && mesh->GetNodes() && !orig_mesh_attr.Size())
+      {
+        BackUpAttributes(*mesh);
+      }
+      int ne = mesh->GetNE();
+      new_mesh = MeshTetToHex(*mesh);
+      Mpi::Print("Added {:d} elements to the mesh during conversion to hexahedra\n",
+                 new_mesh.GetNE() - ne);
+      mesh = &new_mesh;
+    }
+  }
+
+  // Return if no modifications were made.
+  if (mesh == orig_mesh.get())
+  {
+    return;
+  }
+
+  // The previous splitting functions remove curvature information from the new mesh. So, if
+  // needed, we interpolate it onto the new mesh with GSLIB. The topology of the mesh must
+  // be finalized for this to work (assumes FinalizeTopology has been called on the new
+  // mesh).
+  if (preserve_curvature && orig_mesh->GetNodes())
+  {
+    // Prepare to interpolate the grid function for high-order nodes from the old mesh to
+    // the new one. This first sets up the new mesh as a linear mesh and constructs a
+    // separate high-order grid function for storing the interpolated nodes.
+    new_mesh.EnsureNodes();
+    const mfem::GridFunction *nodes = orig_mesh->GetNodes();
+    const mfem::FiniteElementSpace *fespace = nodes->FESpace();
+    mfem::Ordering::Type ordering = fespace->GetOrdering();
+    int order = fespace->GetMaxElementOrder();
+    int sdim = orig_mesh->SpaceDimension();
+    bool discont =
+        (dynamic_cast<const mfem::L2_FECollection *>(fespace->FEColl()) != nullptr);
+    mfem::FiniteElementSpace new_fespace(&new_mesh, fespace->FEColl(), sdim, ordering);
+    mfem::GridFunction new_nodes(&new_fespace);
+
+    mfem::Array<int> vdofs;
+    mfem::Vector vals, elem_vals, xyz;
+    mfem::DenseMatrix pointmat;
+    int start = 0;
+    for (int e = 0; e < orig_mesh->GetNE(); e++)
+    {
+      // Get the high-order nodes restricted to this parent element, always returned
+      // byNODES.
+      fespace->GetElementVDofs(e, vdofs);
+      nodes->GetSubVector(vdofs, vals);
+
+      // Find all child elements of this parent in the split mesh and restore their correct
+      // original attribute.
+      int attr = new_mesh.GetAttribute(start);
+      int end = start;
+      while (end < new_mesh.GetNE() && new_mesh.GetAttribute(end) == attr)
+      {
+        new_mesh.SetAttribute(end++, orig_mesh_attr[e]);
+      }
+
+      // Special case: Parent element is unsplit, no interpolation is needed.
+      if (end == start + 1)
+      {
+        new_fespace.GetElementVDofs(start, vdofs);
+        new_nodes.SetSubVector(vdofs, vals);
+        start = end;
+        continue;
+      }
```

Comment:
In the MakeSimplex and MakeHex methods, neither allow mixed presently, so is this an edge case that can ever be hit? I.e. all tets get converted, and all hexes get converted. Is this just anticipating mixed meshes?